src/main/java/core/GraphReducer.java

```diff
@@ -66,12 +67,54 @@ public static void determineParents(HashMap<Integer, Node> nodeMap) {
             // than the number of nodes after previous reduction.
             if ((previousMapSize - currentMapSize) <= minDelta) {
                 levelMaps.set(i - 1, levelMap);
+                int maxDepth = 20;
+                for (int j = i; maxDepth < 1001; j++) {
+                    HashMap<Integer, Node> levelMap2 = collapse2(levelMaps.get(j - 1), j - 1, maxDepth);
+                    int previousMapSize2 = levelMaps.get(j - 1).size();
+                    int currentMapSize2 = levelMap2.size();
+                    if ((previousMapSize2 - currentMapSize2) == 0) {
+                        levelMaps.set(j - 1, levelMap2);
+                        maxDepth += 5;
+                        j--;
+                    } else {
+                        levelMaps.add(levelMap2);
+                    }
+                }
                 return levelMaps;
             } else {
                 levelMaps.add(levelMap);
             }
         }
+    }
+
+    /**
+     * Reduce the number of nodes in a graph by collapsing vertically and horizontally.
+     *
+     * @param map       A HashMap containing all nodes in the graph.
+     * @param zoomLevel The current zoomLevel
+     * @param maxDepth The maximumDepth allowed for a complexNode
+     * @return A collapsed map.
+     */
+    public static HashMap<Integer, Node> collapse2(HashMap<Integer, Node> map, int zoomLevel, int maxDepth) {
```

Comment:
collapse2 is a really bad name for a method. What is 2nd about this collapse method compared to the other collapse() method?